pkg/virt-controller/services/serialconsolelog.go

```diff
@@ -0,0 +1,91 @@
+package services
+
+import (
+	"fmt"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/util"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+func generateSerialConsoleLogContainer(vmi *v1.VirtualMachineInstance, image string, config *virtconfig.ClusterConfig) *k8sv1.Container {
```

Comment:
You may need a cleanup socket like we have for containerdisks. When the virt-launcher pod goes down, we have for years no good guarantee that all containers in the workload will receive a termination signal.

If not addressed, this can lead to only partially shut-down virt-launcher pods, blocking migrations and reaching a clear Terminated state.

I can also confirm that this is still an issue nowadays, based on various sidecars I have seen in use recently which don't have such a cleanup mechanism.

Reply:
Thanks, I tried enabling this feature by default with another commit and now I'm exactly noticing this partial shutdown behaviour on different CI lanes so the risk is absolutely concrete.

For simplicity I was simply directly executing the system `tail` command as PID 1 in the `guest-console-log` container. I see that `kubevirt/cmd/container-disk-v2alpha/main.c` instead is a custom binary that (on top of waiting for signals) is also periodically checking the existence of a  socket file as an indicator to shutdown itself.

Are you suggesting to re-implement a custom `tail` command that also watches for this socket file or a simple bash script with a background function and some trap logic could potentially be enough?

Reply:
Now I started adding `safe-tail.sh` as a small bash scripts that wraps `tail` and terminates it when the correspondent socket file for the serial console disappears.

Reply:
We don't accept bash scripts anymore. Only go code. They are super hard to maintain and test and are actually most of the time wrong in the details.

Reply:
OK, I implemented something in golang based on github.com/nxadm/tail adding logic to quickly terminate on:
- sigterm
- the unix socket file used for the serial console port disappearing
- a termination file generated by virt-launcher-monitor if previous two mechanisms failed